### ISSUE TITLE
Is it possible to define a mutation operator that takes arguments?

### ISSUE THREAD

**Issue by AndrewC19** [sixty-north/cosmic-ray#528]
I would like to define a custom mutation operator that looks for particular variables in the program-under-test and replaces them with random constants. For example, I might want to replace all uses of a variable named `x1` with some randomly sampled int.

Is there a straightforward way to achieve this using cosmic-ray? I have tried defining my own `Operator` sub-class but I need a way to pass the target variable to the mutation operator.

Thanks in advance.

**Comment by abingham** [comment]
There is not currently a way to do that. I guess broadly speaking you'd want to be able to do some per-operator configuration in the configuration file and have it take effect when you exec your session. This seems totally reasonable. A few things come to mind which we'd need to think about.

1. How will this argument information get communicated through distributors? Right now, the WorkItems that we send over the distributor don't have configuration information in them, so the workers - which could be running anywhere - don't know about it. We could send the complete configuration over to each worker somehow, or - more directly addressing your issue, and I think my preference - each WorkItem could also include some kind of "arguments" struct that the operator could interpret as it sees fit. I don't think this would be too hard, really.

2. Do we need to support multiple sets of arguments for any given operators? That is, suppose you wanted to replace all variables named `x1` *and* you wanted to replace all  variables named `x2`? Do operators now become templates (in the C++ sense, sorta) where each set of arguments actually generates a new independent operator? Something like this could possibly be handled by our `OperatorProvider` system, but the providers would probably need to be handed the configuration. (and perhaps I'm just overthinking this).

There are probably other angles to this that I'm not seeing right now. Does any of this sound reasonable to you, or along the lines of what you had in mind? I don't know when I'd be able to work on this, but it feels like a good "airplane project"...I just need a long trip somewhere!

**Comment by AndrewC19** [comment]
Thanks for the quick response @abingham.

Both of these solutions sound sensible to me but perhaps some additional info will help to narrow it down.

Ultimately, I want to be able to make three kinds of mutation that target specific lines of code based on their variable usage. Here I will use the term 'used' to mean that a variable appears on the right hand side of `=` and 'declared' for variables on the left hand side (e.g. in `y = 2*x1`, `y` is declared and `x1` is used). 

1. Find lines of code where a variable, say `x1`, is used and replace `x1` with a constant (e.g. `y1 = 2*x1` --> `y1 = 2*5`).
2. Find lines of code where a variable, say `y1`, is declared *and* a second variable, say `x1`, is used and replace `x1` with a constant (e.g. `y1 = 2*x1 + x2` --> `y1=2*10 + x2`).
3. Find lines of code where a variable, say `y1`, is declared and introduce a usage of a second variable, say `x1` (e.g. `y1 = 2*x2` --> `y1 = 2*x2 + x1`).

The first mutation is effectively removing all effects of the specific variable (`x1`). The second mutation removes all effects of one variable (`x1`) on another (`y1`)  (i.e. a refinement of the first mutation). The third mutation is adding an effect of one variable (`x1) on another (`y1`).

It's worth noting that each of these mutations could be achieved in different ways. For example, to remove the effect of `x1` on `y1` we could also delete usages of `x1` in statements of `y1` rather than replacing with a constant. The main blocker at the moment, however, is the inability to target specific variables.

In my opinion, since the `mutation_positions` method of `Operator` handles the identification of sutiable mutation points, it would make sense that this method handles targetting of specific variables. Of course, I am not sure how easy/suitable such a change would be as I am not familiar with the infrastructure behind cosmic-ray.

Hope that helps!

**Comment by AndrewC19** [comment]
To follow up on this, I have implemented a `VariableReplacement` operator that is capable of mutations 1 and 2 of the above post (i.e. replacing all usages of a named variable and replacing usages of a named variable in the declaration of a second named variable).

```python
"""Implementation of the variable-replacement operator."""
from .operator import Operator
from parso.python.tree import Name, Number
from random import randint


class VariableReplacer(Operator):
    """An operator that replaces usages of named variables."""

    def __init__(self, cause_variable, effect_variable=None):
        self.cause_variable = cause_variable
        self.effect_variable = effect_variable

    def mutation_positions(self, node):
        """Mutate usages of the specified cause variable. If an effect variable is also
        specified, then only mutate usages of the cause variable in definitions of the
        effect variable."""

        if isinstance(node, Name) and node.value == self.cause_variable:

            # Confirm that name node is used on right hand side of the expression
            expr_node = node.search_ancestor('expr_stmt')
            if expr_node:
                cause_variables = expr_node.get_rhs().children
                if node in cause_variables:
                    mutation_position = (node.start_pos, node.end_pos)

                    # If an effect variable is specified, confirm that it appears on left hand
                    # side of the expression
                    if self.effect_variable:
                        effect_variable_names = [v.value for v in expr_node.get_defined_names()]
                        if self.effect_variable in effect_variable_names:
                            yield mutation_position

                    # If no effect variable is specified, any occurrence of the cause variable
                    # on the right hand side of an expression can be mutated
                    else:
                        yield mutation_position

    def mutate(self, node, index):
        """Replace cause variable with random constant."""
        assert isinstance(node, Name)

        return Number(start_pos=node.start_pos, value=str(randint(-100, 100)))

    @classmethod
    def examples(cls):
        return (
            # for cause_variable='x'
            ('y = x + z', 'y = 10 + z'),
            # for cause_variable='x' and effect_variable='y'
            ('j = x + z\ny = x + z', 'j = x + z\ny = -2 + z'),
            # for cause_variable='x' and effect_variable='j',
            ('j = x + z\ny = x + z', 'j = 1 + z\ny = -2 + z'),
            # for cause_variable='x'
            ('y = 2*x + 10 + j + x**2', 'y=2*10 + 10 + j + -4**2'),
        )
```

The class works if I manually modify `cosmic_ray/commands/init.py` and `src/cosmic_ray/mutating.py` to instantiate the operator with arguments, and creates mutations such as (where `cause_variable='x'` and `effect_variable='y'`:

```python
--- mutation diff ---
--- acalculator.py
+++ bcalculator.py
@@ -1,5 +1,5 @@
 def mul(x, z):
     j = x * z
-    y = x * j
+    y =28 * j
     return y
```

In short, all that remains is a systematic way to pass arguments to the points where the operator is instantiated. Any thoughts?
